rtmr/rtmr.go

```diff
@@ -0,0 +1,210 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package rtmr provides an API to the configfs/tsm/rtmr subsystem for
+// extending runtime measurements to RTMR registers.
+package rtmr
+
+import (
+	"crypto"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"github.com/google/go-configfs-tsm/configfs/configfsi"
+)
+
+const (
+	rtmrSubsystem = "rtmrs"
+	tsmRtmrPrefix = configfsi.TsmPrefix + "/" + rtmrSubsystem
+	// The digest of the rtmr register.
+	tsmRtmrDigest = "digest"
+	// A Runtime Measurement Register (RTMR) hardware index.
+	tsmPathIndex = "index"
+	// A representation of the architecturally defined mapping between this RTMR and one or more TCG TPM PCRs
+	tsmPathTcgMap = "tcg_map"
+)
+
+// Extend is a struct that represents a rtmr entry in the configfs.
+type Extend struct {
+	RtmrIndex int
+	entry     *configfsi.TsmPath
+	client    configfsi.Client
+}
+
+// Response is a struct that represents the response of reading a rtmr entry in the configfs.
+type Response struct {
+	RtmrIndex int
+	digest    []byte
+	tcgMap    []byte
+}
+
+func (r *Extend) attribute(subtree string) string {
+	a := *r.entry
+	a.Attribute = subtree
+	return a.String()
+}
+
+// extendDigest extends the measurement to the rtmr with the given hash.
+func (r *Extend) extendDigest(hash []byte) error {
+	if err := r.client.WriteFile(r.attribute(tsmRtmrDigest), hash); err != nil {
+		return fmt.Errorf("could not write digest to rmtr%d: %v", r.RtmrIndex, err)
+	}
+	return nil
+}
+
+// getDigest returns the digest of the rtmr.
+func (r *Extend) getDigest() ([]byte, error) {
+	return r.client.ReadFile(r.attribute(tsmRtmrDigest))
+}
+
+// getTcgMap returns the tcg map of the rtmr.
+func (r *Extend) getTcgMap() ([]byte, error) {
+	return r.client.ReadFile(r.attribute(tsmPathTcgMap))
+}
+
+// validateIndex checks if the rtmr index is valid.
+func (r *Extend) validateIndex() error {
+	indexBytes, err := r.client.ReadFile(r.attribute(tsmPathIndex))
+	if err != nil {
+		return fmt.Errorf("could not read index %s: %v", r.attribute(tsmPathIndex), err)
+	}
+	index, err := configfsi.Kstrtouint(indexBytes, 10, 64)
+	if err != nil {
+		return fmt.Errorf("could not convert index %s to integer: %v", indexBytes, err)
+	}
+	if (int)(index) != r.RtmrIndex {
+		return fmt.Errorf("rtmr index %d does not match the expected index %d", index, r.RtmrIndex)
```

Comment:
Acknowledged.
`validateIndex` is a non public method and this error does not return to the public function.

Reply:
Then you're not using any of these constructed strings. Either surface the errors or collapse to bool for validateIndex and make searchRtmrInterface just return *Extend only and just nil if it can't find anything.

Reply:
Make sense. Done